fuel-derive/src/deserialize.rs

```diff
@@ -0,0 +1,211 @@
+            *buffer = buf; // Restore buffer position
+        }
+    } else {
+        quote! {
+            let raw_discr = <::core::primitive::u64 as ::fuel_types::canonical::Deserialize>::decode(buffer)?;
+        }
+    };
+
+    // Handle #[canonical(discriminant = Type)]
+    let mapped_discr = if let Some(discr_type) =
+        attrs.discriminant.or(attrs.inner_discriminant)
+    {
+        quote! { {
+            use ::num_enum::{TryFromPrimitive, IntoPrimitive};
+            let Ok(discr) = #discr_type::try_from_primitive(raw_discr) else {
+                return ::core::result::Result::Err(::fuel_types::canonical::Error::UnknownDiscriminant)
+            };
+            discr
+        } }
+    } else {
+        quote! { raw_discr }
+    };
+
+    // Handle #[canonical(deserialize_with = function)]
+    if let Some(helper) = attrs.deserialize_with {
+        return s.gen_impl(quote! {
+            gen impl ::fuel_types::canonical::Deserialize for @Self {
+                fn decode_static<I: ::fuel_types::canonical::Input + ?Sized>(buffer: &mut I) -> ::core::result::Result<Self, ::fuel_types::canonical::Error> {
+                    let raw_discr = <::core::primitive::u64 as ::fuel_types::canonical::Deserialize>::decode(buffer)?;
+                    #helper(#mapped_discr, buffer)
+                }
+
+                fn decode_dynamic<I: ::fuel_types::canonical::Input + ?Sized>(&mut self, buffer: &mut I) -> ::core::result::Result<(), ::fuel_types::canonical::Error> {
+                    ::core::result::Result::Ok(())
+                }
+            }
+        })
+    }
+
+    s.gen_impl(quote! {
+        gen impl ::fuel_types::canonical::Deserialize for @Self {
+            fn decode_static<I: ::fuel_types::canonical::Input + ?Sized>(buffer: &mut I) -> ::core::result::Result<Self, ::fuel_types::canonical::Error> {
+                #decode_discriminant
+                match #mapped_discr {
+                    #decode_static
+                    _ => ::core::result::Result::Err(::fuel_types::canonical::Error::UnknownDiscriminant),
+                }
+            }
+
+            fn decode_dynamic<I: ::fuel_types::canonical::Input + ?Sized>(&mut self, buffer: &mut I) -> ::core::result::Result<(), ::fuel_types::canonical::Error> {
+                match self {
+                    #(
+                        #decode_dynamic
+                    )*
+                    _ => return ::core::result::Result::Err(::fuel_types::canonical::Error::UnknownDiscriminant),
+                };
+
+                ::core::result::Result::Ok(())
+            }
+        }
+    })
+}
+
+/// Derives `Deserialize` trait for the given `struct` or `enum`.
+pub fn deserialize_derive(mut s: synstructure::Structure) -> TokenStream2 {
+    s.bind_with(|_| synstructure::BindStyle::RefMut)
+        .add_bounds(synstructure::AddBounds::Fields)
+        .underscore_const(true);
+
+    match s.ast().data {
+        syn::Data::Struct(_) => deserialize_struct(&mut s),
```

Comment:
Does this automatically support tuple structs as well?

Reply:
Yes. See syn documentation here: https://docs.rs/syn/latest/syn/enum.Data.html